sshd-common/src/main/java/org/apache/sshd/common/config/keys/OpenSshCertificate.java

```diff
@@ -91,7 +104,83 @@
      */
     static boolean isValidNow(OpenSshCertificate cert) {
         long now = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
-        return Long.compareUnsigned(cert.getValidAfter(), now) <= 0
-                && Long.compareUnsigned(now, cert.getValidBefore()) < 0;
+        return Long.compareUnsigned(cert.getValidAfterEpochSeconds(), now) <= 0
+                && Long.compareUnsigned(now, cert.getValidBeforeEpochSeconds()) < 0;
+    }
+
+    /**
+     * According to <A HREF=
+     * "https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.certkeys#L222-L262">PROTOCOL.certkeys</A>:
+     *
+     * Critical Options is a set of bytes that is
+     *
+     * [overall length][name(string)][data(string)]...
+     *
+     * Where each Critical Option is encoded as a name (string) and data (string)
+     *
+     * Then the entire name + data strings are added as bytes (which will get a length prefix)
+     */
+    class CriticalOption {
+
+        private String name;
+        private String data;
+
+        public CriticalOption(String name, String data) {
+            this.name = name;
+            this.data = data;
+        }
+
+        public CriticalOption(String name) {
+            this.name = name;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
+
+        public String getData() {
+            return data;
+        }
+
+        public void setData(String data) {
+            this.data = data;
+        }
+
+        @Override
+        public String toString() {
+            return "CriticalOption{" +
+                   "name='" + name + '\'' +
+                   ", data='" + data + '\'' +
+                   '}';
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            CriticalOption that = (CriticalOption) o;
+
+            if (name != null ? !name.equals(that.name) : that.name != null) {
+                return false;
+            }
+            return data != null ? data.equals(that.data) : that.data == null;
```

Comment:
1. Using Objects.equals() and Objects.hash() in hashCode() below would simplify this; no need to check for null.
2. equals and hashCode based on non-final fields is not a good idea. If a CriticalOption object is stored in a HashMap and then name or data is modified, you'll never find it again.

I'd make the fields final and drop the setters.